File: src/main/java/com/soapdogg/lispInterpreter/asserter/AtomRootNodeAsserter.java
package com.soapdogg.lispInterpreter.asserter;

import com.soapdogg.lispInterpreter.comparator.NodeValueComparator;
import com.soapdogg.lispInterpreter.datamodels.AtomNode;
import com.soapdogg.lispInterpreter.determiner.NumericStringDeterminer;
import com.soapdogg.lispInterpreter.exceptions.NotAtomicException;
import lombok.AllArgsConstructor;

@AllArgsConstructor(staticName = "newInstance")
public class AtomRootNodeAsserter {

    private final NumericStringDeterminer numericStringDeterminer;
    private final NodeValueComparator nodeValueComparator;

    public void assertAtomRootNode(
        final AtomNode atomNode
    ) throws NotAtomicException {
        boolean isNotNumeric = !numericStringDeterminer.isStringNumeric(atomNode.getValue());
        boolean isNotT = !nodeValueComparator.equalsT(atomNode.getValue());
        boolean isNotNil = !nodeValueComparator.equalsNil(atomNode.getValue());
        if (isNotNumeric && isNotT && isNotNil){
            throw new NotAtomicException("Error! " + atomNode.getValue() + " is not a valid atomic value!\n");
        }
    }
}

File: src/main/java/com/soapdogg/lispInterpreter/asserter/AtomRootNodeAsserter.kt
package com.soapdogg.lispInterpreter.asserter

import com.soapdogg.lispInterpreter.constants.ReservedValuesConstants
import com.soapdogg.lispInterpreter.datamodels.AtomNode
import com.soapdogg.lispInterpreter.determiner.NumericStringDeterminer
import com.soapdogg.lispInterpreter.exceptions.NotAtomicException

class AtomRootNodeAsserter(
    private val numericStringDeterminer: NumericStringDeterminer
) {

    fun assertAtomRootNode(
        atomNode: AtomNode
    ) {
        val isNotNumeric = !numericStringDeterminer.isStringNumeric(atomNode.value)
        val isNotT = atomNode.value != ReservedValuesConstants.T
        val isNotNil = atomNode.value != ReservedValuesConstants.NIL
        if (isNotNumeric && isNotT && isNotNil) {
            throw NotAtomicException("""Error! ${atomNode.value} is not a valid atomic value!${'\n'}""")
        }
    }
}